frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -504,79 +520,153 @@ static void checkForParenlessMethodFieldRedefinition(Context* context,
   }
 }
 
-static const TypedFnSignature* const&
-typedSignatureInitialQuery(Context* context,
-                           const UntypedFnSignature* untypedSig) {
-  QUERY_BEGIN(typedSignatureInitialQuery, context, untypedSig);
+static const std::map<ID, std::pair<QualifiedType, std::vector<ID>>>&
+signatureOuterVariablesQuery(Context* context,
+                             const TypedFnSignature* sig,
+                             const PoiScope* poiScope) {
+  QUERY_BEGIN(signatureOuterVariablesQuery, context, sig, poiScope);
+  CHPL_ASSERT(false && "Should be set elsewhere!");
+  std::map<ID, std::pair<QualifiedType, std::vector<ID>>> ret;
+  return QUERY_END(ret);
+}
 
+static const std::map<ID, std::pair<QualifiedType, std::vector<ID>>>*
+signatureOuterVariables(Context* context, const TypedFnSignature* sig,
+                        const PoiScope* poiScope) {
+  constexpr auto f = signatureOuterVariablesQuery;
+  if (!context->hasCurrentResultForQuery(f, { sig, poiScope })) return nullptr;
+  return &f(context, sig, poiScope);
+}
+
+static const TypedFnSignature*
+typedSignatureInitialImpl(Context* context,
+                          const UntypedFnSignature* untypedSig,
+                          const CallerDetails& caller) {
   const TypedFnSignature* result = nullptr;
   const AstNode* ast = parsing::idToAst(context, untypedSig->id());
   const Function* fn = ast->toFunction();
 
-  if (fn != nullptr) {
-    // look at the parent scopes to find the parent function, if any
-    const UntypedFnSignature* parentFnUntyped = nullptr;
-    const TypedFnSignature* parentFnTyped = nullptr;
-    ID parentFnId = parentFunctionId(context, fn->id());
-    if (!parentFnId.isEmpty()) {
-      auto parentAst = parsing::idToAst(context, parentFnId);
-      auto parentFn = parentAst->toFunction();
-      parentFnUntyped = UntypedFnSignature::get(context, parentFn);
-      parentFnTyped = typedSignatureInitial(context, parentFnUntyped);
+  if (fn == nullptr) return nullptr;
+
+  // If it is a nested function, then the parent signature should be stored
+  // in the 'CallerDetails'. If there are no details, then try to call the
+  // 'typedSignatureInitial', but it will give up if the signature for the
+  // "parent's parent" is generic or if the parent contains outer variables.
+  const TypedFnSignature* parentSignature = nullptr;
```

Comment:
IMO `UntypedFnSignature` should have a field indicating it is a nested function & a field indicating if it's a function containing nested functions. That should make it easy to avoid these computations in the common case that there is no nesting.

Reply:
I'm not sure if we want to store those details inside `UntypedFnSignature`, perhaps instead I could add a separate query later to compute that?

Reply:
Why not? They should be syntactically apparent. `UntypedFnSignature` already has a bunch of `bool`s so it should be possible to add it without taking any more space.